### PR TITLE
Distinct

### PR DESCRIPTION
**Pull request by jsegaran**
This is the distinct operator, but at the moment it requires full materialization to work at the moment, partial isn't yet supported for this operator.

### CODE REVIEW

File: dataflow/src/ops/distinct.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
+use std::collections::HashMap;
+
+use prelude::*;
+
+/// This will get Distinct elements from each group
+/// TODO
+
```
Comment (jonhoo) [inline dataflow/src/ops/distinct.rs]: It' be great if you could expand this comment to explain what the operator does, how, and what it's limitations are :) In Rust, you also have to not leave a space between a documentation comment and the item it is documenting (i.e., the newline above has to go away).

File: dataflow/src/ops/distinct.rs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+pub struct Distinct {
+    // Parent Node
+    src: IndexPair,
+
+    us: Option<IndexPair>,
+
+    group_by: Vec<usize>,
+
+    // A list of records that have been seen
+    seen: HashMap<Vec<DataType>, bool>,
```
Comment (jonhoo) [inline dataflow/src/ops/distinct.rs]: It doesn't look like this field is currently used anywhere? Was the idea to use this instead of the
```
let mut seen = HashMap::new();
```
inside `on_input`?

File: dataflow/src/ops/distinct.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+impl Distinct {
+    pub fn new(
+        src: NodeIndex,
+        group_by: Vec<usize>,
+        ) -> Self {
+
+        // TODO add implementation for group by
```
Comment (jonhoo) [inline dataflow/src/ops/distinct.rs]: Not sure what this `TODO` refers to?

File: dataflow/src/ops/distinct.rs [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+impl Ingredient for Distinct {
+    /// Returns a clone of the node
+    fn take(&mut self) -> NodeOperator {
+        Clone::clone(self).into()
+    }
+
+    fn ancestors(&self) -> Vec<NodeIndex> {
+        // I think this is right?
```
Comment (jonhoo) [inline dataflow/src/ops/distinct.rs]: Yup, it is!

File: dataflow/src/ops/distinct.rs (hunk 2 above)
Comment (jonhoo) [reply dataflow/src/ops/distinct.rs]: It's also likely a good idea to document that these do not actually hold *all* records that have been seen, just the ones that have been seen in *this* packet. That is, they are cleared for every input.

File: dataflow/src/ops/distinct.rs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+            if seen.contains_key(&group){
+                if !positive{
+                    seen.remove(&group);
+                    cleaned.push(rec.clone());
+                }
+            } else {
+                cleaned.push(rec.clone());
+                seen.insert(group.clone(), true);
+            }
+        }
```
Comment (jonhoo) [inline dataflow/src/ops/distinct.rs]: I think you could actually make this code a fair amount nicer by just sorting the batch first and then walking it in order. That way you'd know to not skip records whenever the key changes. You can take a look at how we do this [for grouped operators](https://github.com/mit-pdos/distributary/blob/67514028f32d94e78ce1ed2513fb963d94083281/dataflow/src/ops/grouped/mod.rs#L162L174). The implementation for [TopK](https://github.com/mit-pdos/distributary/blob/67514028f32d94e78ce1ed2513fb963d94083281/dataflow/src/ops/topk.rs#L131) is even closer to what you want to do; it also detects where groups start and stop, but the code is a little hairier as a result. Changing the code that way would also eliminate both `seen` and `cleaned` I think.

File: dataflow/src/ops/distinct.rs [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+                    if self.group_by.iter().any(|col| col == &i) {
+                        Some(v)
+                    } else {
+                        None
+                    }
+                })
+                .cloned()
+                .collect::<Vec<_>>();
+            let positive = rec.is_positive();
+            if positive {
```
Comment (jonhoo) [inline dataflow/src/ops/distinct.rs]: I'd probably put this `if` deeper inside so that we avoid duplicating the `db.lookup` code that is the same across both conditions.

File: dataflow/src/ops/distinct.rs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                match db.lookup(group_by, &KeyType::from(&group[..])) {
+                    LookupResult::Some(rr) => {
+                        //println!("record {:?}", rr);
+                        if rr.len() == 0 {
+                            output.push(rec.clone());
+                        }
+                    },
+                    LookupResult::Missing => {
```
Comment (jonhoo) [inline dataflow/src/ops/distinct.rs]: I would put `unimplemented!("Distinct does not yet support partial")` here so we don't miss (heh) this case should it happen due to a bug.

File: dataflow/src/ops/mod.rs [hunk 8]
```diff
@@ -81,6 +84,8 @@ macro_rules! impl_ingredient_fn_mut {
             NodeOperator::TopK(ref mut i) => i.$fn($($arg),*),
             NodeOperator::Trigger(ref mut i) => i.$fn($($arg),*),
             NodeOperator::Rewrite(ref mut i) => i.$fn($($arg),*),
+            NodeOperator::Distinct(ref mut i) => i.$fn($($arg),*),
+
```
Comment (jonhoo) [inline dataflow/src/ops/mod.rs]: Spurious whitespace.

File: examples/basic-recipe.rs [hunk 9]
```diff
@@ -30,8 +30,9 @@ fn main() {
 
     // set up Soup via recipe
     let mut builder = ControllerBuilder::default();
-    builder.log_with(distributary::logger_pls());
+    //builder.log_with(distributary::logger_pls());
     builder.set_worker_threads(2);
+    builder.disable_partial();
```
Comment (jonhoo) [inline examples/basic-recipe.rs]: We should probably not change the provided example code this way :) Maybe make a test that executes this, and leave the example code the way it was?

File: mir/src/node.rs [hunk 10]
```diff
@@ -871,7 +893,7 @@ impl Debug for MirNodeType {
                     .join(" ⋃ ");
 
                 write!(f, "{}", cols)
-            }
+            },
```
Comment (jonhoo) [inline mir/src/node.rs]: Spurious change.

File: src/controller/mir_to_flow.rs [hunk 11]
```diff
@@ -739,6 +752,35 @@ pub(crate) fn make_project_node(
...
+    columns: &[Column],
+    group_by: &Vec<Column>,
+    mig: &mut Migration,
+) -> FlowNode {
+    let parent_na = parent.borrow().flow_node_addr().unwrap();
+    let column_names = column_names(columns);
+
+    let group_by_indx = if group_by.is_empty() {
+        // no query parameters, so we index on the first column
+        vec![0 as usize]
```
Comment (jonhoo) [inline src/controller/mir_to_flow.rs]: Shouldn't the default be to be distinct by *all* columns?

File: src/controller/sql/mir/grouped.rs [hunk 12]
```diff
@@ -204,16 +204,16 @@ pub fn make_grouped(
...
                     group_cols.iter().collect(),
                     parent_node,
                 );
 
-                *prev_node = Some(n.clone());
-                func_nodes.push(n);
-                node_count += 1;
+                *prev_node = Some(nodes.last().unwrap().clone());
+                node_count += nodes.len();
+                func_nodes.extend(nodes);
```
Comment (jonhoo) [inline src/controller/sql/mir/grouped.rs]: This change seems unrelated to distinct?

**Comment by jonhoo** [comment]
Nice! Congratulations on your first Soup operator :D I've left some comments inline, but overall looks pretty good. I'll leave the MIR changes for @ms705 to look at.

File: src/controller/mir_to_flow.rs (hunk 11 above)
Comment (ms705) [reply src/controller/mir_to_flow.rs]: Agree, all columns should be the default here.

File: src/controller/sql/mir/mod.rs [hunk 13]
```diff
@@ -805,19 +807,45 @@ impl SqlToMirConverter {
...
+                        name: a.clone(),
+                        alias: None,
+                        table: func_col.table.clone(),
+                        function: func_col.function.clone(),
+                    },
+                };
+                let mut dist_col = Vec::new();
+                dist_col.push(over);
+                dist_col.extend(group_cols.clone());
+                //println!("dist_cols {:?}", dist_col);
```
Comment (ms705) [inline src/controller/sql/mir/mod.rs]: This can be dropped now, I think.

File: src/controller/sql/mir/mod.rs [hunk 14]
```diff
@@ -805,19 +807,45 @@ impl SqlToMirConverter {
...
 
-        let mknode = |over: &Column, t: GroupedNodeType| {
-            self.make_grouped_node(name, &func_col, (parent, &over), group_cols, t)
+        let mut out_nodes = Vec::new();
+
+        let mknode = |over: &Column, t: GroupedNodeType, d: bool| {
+            if d {
+                let new_name = name.clone().to_owned() + "distinct";
+
+                let computed_col = match func_col.alias {
```
Comment (ms705) [inline src/controller/sql/mir/mod.rs]: I think `computed_col` is not needed here, or is it?

File: src/controller/sql/mir/mod.rs [hunk 15]
```diff
@@ -1581,6 +1629,22 @@ impl SqlToMirConverter {
                     new_node_count += 1;
                 }
 
+                info!(self.log, "distinct defined as {}", st.distinct);
```
Comment (ms705) [inline src/controller/sql/mir/mod.rs]: I don't think we need this at `INFO` log level; drop or reduce?

File: src/controller/sql/mir/mod.rs [hunk 16]
```diff
@@ -1581,6 +1629,22 @@ impl SqlToMirConverter {
...
+                if st.distinct {
+                    info!(self.log, "Adding distinct node to the query, distinct is set");
```
Comment (ms705) [inline src/controller/sql/mir/mod.rs]: As above.

**Comment by ms705** [comment]
Great work! :+1: 

MIR changes look good to me, modulo minor comments.

It'd be good to get this merged, since I plan to rewrite some of the MIR code that this PR modifies. @jsegaran are you likely to have time to finish it off? If not, @jonhoo or myself can take on rebasing, making the changes and merging it.

**Comment by ms705** [comment]
Thanks for rebasing! I think this is now good to merge -- @jonhoo?

**Comment by ms705** [comment]
Merge -- good work, @jsegaran!
